scio-core/src/main/scala/com/spotify/scio/util/BloomFilter.scala

```diff
@@ -0,0 +1,600 @@
+
+  def +=(other: A): MutableBF[A]
+  // scalastyle:on method.name
+
+  def checkAndAdd(item: A): (MutableBF[A], ApproximateBoolean)
+
+  def contains(item: A): ApproximateBoolean =
+    if (maybeContains(item)) {
+      // The false positive probability (the probability that the Bloom filter erroneously
+      // claims that an element x is in the set when x is not) is roughly
+      // p = (1 - e^(-numHashes * setCardinality / width))^numHashes
+      // See: http://en.wikipedia.org/wiki/Bloom_filter#Probability_of_false_positives
+      //
+      // However, the true set cardinality may not be known. From empirical evidence, though,
+      // it is upper bounded with high probability by 1.1 * estimatedCardinality (as long as the
+      // Bloom filter is not too full), so we plug this into the formula instead.
+      // TODO: investigate this upper bound and density more closely (or derive a better formula).
+      // TODO: The following logic is same for immutable Bloom Filters and may be referred here.
+      val fpProb =
+        if (density > 0.95) {
+          1.0 // No confidence in the upper bound on cardinality.
+        } else {
+          math.pow(1 - scala.math.exp(-numHashes * size.estimate * 1.1 / width), numHashes)
+        }
+
+      ApproximateBoolean(isTrue = true, 1 - fpProb)
+    } else {
+      // False negatives are not possible.
+      ApproximateBoolean.exactFalse
+    }
+
+  /**
+   * This may be faster if you don't care about evaluating
+   * the false positive probability
+   */
+  def maybeContains(item: A): Boolean
+
+  // Estimates the cardinality of the set of elements that have been
+  // inserted into the Bloom Filter.
+  def size: Approximate[Long]
+
+  def toBitSet: util.BitSet
+
+  def copy: MutableBF[A]
+
+  /**
+   * Compute the Hamming distance between the two Bloom filters
+   * `a` and `b`. The distance is defined as the number of bits that
+   * need to change to in order to transform one filter into the other.
+   * This is computed using XOR but it doesn't mutate any BloomFilters
+   */
+  def hammingDistance(that: MutableBF[A]): Int =
+    (this, that) match {
+      // Comparing with empty filter should give number
+      // of bits in other set
+      case (x: MutableBFZero[A], y: MutableBFZero[A]) => 0
+      case (x: MutableBFZero[A], y: MutableBF[A])     => y.numBits
+      case (x: MutableBF[A], y: MutableBFZero[A])     => x.numBits
+
+      // Otherwise compare as bit sets
+      case _ =>
+        // hammingDistance should not mutate BloomFilter
+        val thisCopy = this.toBitSet.clone().asInstanceOf[util.BitSet]
+        thisCopy.xor(that.toBitSet)
+        thisCopy.cardinality()
+    }
+
+}
+
+/**
+ * Empty bloom filter.
+ */
+private[scio] final case class MutableBFZero[A](hashes: KirMit32Hash[A]) extends MutableBF[A] {
+
+  def toBitSet: util.BitSet = new util.BitSet()
+
+  def numHashes: Int = hashes.numHashes
+
+  def width: Int = hashes.width
+
+  def numBits: Int = 0
```

Comment:
Hi @anish749 you could have kept all of this as `val`. the important part was to remove the `lazy`

Reply:
umm, I was using `def` first then moved to `lazy val` to not re-evaluate. I am a bit hesitant with `val` because it would be allocating values which are copies of other `Int` and it might increase memory footprint. Also in most cases here, they are not called (hence `lazy val` didn't improve speed compared to `def`)

Reply:
It's a valid concern, however it seems to me that some of them will be re-evaluated multiple times and I'm not sure if this cost is lower compared to the usage of `val` given that some of them are actually just ref's and also that there's higher change of getting inlining by the VM. Also note that you were comparing with `lazy val` and it has hidden cost as well.

Another thing is that `val` conveys a better meaning since these values will not actually change.

This interesting stuff and It would be awesome if we could bench it 😄

Reply:
on a second thought, the usage of `val` here would need to be carefully applied, because it can really increase object size. Let's keep this as is for now! numbers look pretty good already!

Reply:
Yes, it's a bit of a trade off.. I did try different approaches, but the jmh benchmarks doesn't give very accurate results on these changes, so its harder to benchmark.